File: src/contexts/cases/infra/schema.py
"""
Cases Context: SQLAlchemy Core Schema

Table definitions for the Cases bounded context using SQLAlchemy Core.
Defines cas_case, cas_attribute, and cas_source_link tables.

These tables use the 'cas_' prefix to identify them as belonging to
the Cases bounded context.
"""

from sqlalchemy import (
    Column,
    DateTime,
    Index,
    Integer,
    MetaData,
    String,
    Table,
    Text,
)

# Metadata for Cases context tables
metadata = MetaData()

# ============================================================
# Table Definitions
# ============================================================

# cas_case - Research units (participants, sites, etc.)
cas_case = Table(
    "cas_case",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("name", String(255), nullable=False),
    Column("description", Text),
    Column("memo", Text),
    Column("owner", String(100)),
    Column("created_at", DateTime),
    Column("updated_at", DateTime),
    Index("idx_cas_case_name", "name"),
)

# cas_attribute - Demographic/categorical data for cases
cas_attribute = Table(
    "cas_attribute",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("case_id", Integer, nullable=False),
    Column("name", String(100), nullable=False),
    Column("attr_type", String(20), nullable=False),  # text, number, date, boolean
    Column("value_text", Text),
    Column("value_number", Integer),
    Column("value_date", DateTime),
    Index("idx_cas_attr_case_name", "case_id", "name", unique=True),
)

# cas_source_link - Case-to-Source associations
# Contains denormalized source_name for cross-context display
cas_source_link = Table(
    "cas_source_link",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("case_id", Integer, nullable=False),
    Column("source_id", Integer, nullable=False),  # References src_source.id
    Column("owner", String(100)),
    Column("date", String(50)),
    Column("source_name", String(255)),  # Denormalized for display
    Index("idx_cas_source_link_unique", "case_id", "source_id", unique=True),
)


def create_all(engine) -> None:
    """
    Create all tables for the Cases context.

    Args:
        engine: SQLAlchemy engine instance
    """
    metadata.create_all(engine)


def drop_all(engine) -> None:
    """
    Drop all Cases context tables (for testing).

    Args:
        engine: SQLAlchemy engine instance
    """
    metadata.drop_all(engine)

File: convex/schema.ts
/**
 * QualCoder Convex Database Schema
 *
 * Defines all tables mirroring the SQLite schema structure.
 * Uses bounded context prefixes for table isolation.
 */

import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  // =========================================================================
  // Coding Context Tables (cod_)
  // =========================================================================

  cod_category: defineTable({
    name: v.string(),
    memo: v.optional(v.string()),
    owner: v.optional(v.string()),
    date: v.optional(v.string()),
    supercatid: v.optional(v.id("cod_category")),
  })
    .index("by_name", ["name"])
    .index("by_parent", ["supercatid"]),

  cod_code: defineTable({
    name: v.string(),
    color: v.optional(v.string()),
    memo: v.optional(v.string()),
    owner: v.optional(v.string()),
    date: v.optional(v.string()),
    catid: v.optional(v.id("cod_category")),
  })
    .index("by_name", ["name"])
    .index("by_category", ["catid"]),

  cod_segment: defineTable({
    cid: v.id("cod_code"),
    fid: v.id("src_source"),
    pos0: v.number(),
    pos1: v.number(),
    seltext: v.optional(v.string()),
    memo: v.optional(v.string()),
    owner: v.optional(v.string()),
    date: v.optional(v.string()),
    important: v.optional(v.number()),
    source_name: v.optional(v.string()),
  })
    .index("by_code", ["cid"])
    .index("by_source", ["fid"])
    .index("by_source_and_code", ["fid", "cid"]),

  // =========================================================================
  // Sources Context Tables (src_)
  // =========================================================================

  src_folder: defineTable({
    name: v.string(),
    parent_id: v.optional(v.id("src_folder")),
    created_at: v.optional(v.string()),
  })
    .index("by_name", ["name"])
    .index("by_parent", ["parent_id"]),

  src_source: defineTable({
    name: v.string(),
    fulltext: v.optional(v.string()),
    mediapath: v.optional(v.string()),
    memo: v.optional(v.string()),
    owner: v.optional(v.string()),
    date: v.optional(v.string()),
    av_text_id: v.optional(v.number()),
    risession: v.optional(v.number()),
    source_type: v.optional(v.string()),
    status: v.optional(v.string()),
    file_size: v.optional(v.number()),
    origin: v.optional(v.string()),
    folder_id: v.optional(v.id("src_folder")),
  })
    .index("by_name", ["name"])
    .index("by_type", ["source_type"])
    .index("by_folder", ["folder_id"])
    .index("by_status", ["status"]),

  // =========================================================================
  // Cases Context Tables (cas_)
  // =========================================================================

  cas_case: defineTable({
    name: v.string(),
    description: v.optional(v.string()),
    memo: v.optional(v.string()),
    owner: v.optional(v.string()),
    created_at: v.optional(v.string()),
    updated_at: v.optional(v.string()),
  }).index("by_name", ["name"]),

  cas_attribute: defineTable({
    case_id: v.id("cas_case"),
    name: v.string(),
    attr_type: v.string(),
    value_text: v.optional(v.string()),
    value_number: v.optional(v.number()),
    value_date: v.optional(v.string()),
  }).index("by_case_and_name", ["case_id", "name"]),

  cas_source_link: defineTable({
    case_id: v.id("cas_case"),
    source_id: v.id("src_source"),
    owner: v.optional(v.string()),
    date: v.optional(v.string()),
    source_name: v.optional(v.string()),
  }).index("by_case_and_source", ["case_id", "source_id"]),

  // =========================================================================
  // Projects Context Tables (prj_)
  // =========================================================================

  prj_settings: defineTable({
    key: v.string(),
    value: v.optional(v.string()),
    updated_at: v.optional(v.string()),
  }).index("by_key", ["key"]),
});
